jerry-ext/args/args.c

```diff
@@ -0,0 +1,97 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript-ext/args.h"
+#include "jerryscript.h"
+
+/**
+ * It validate the JS arguments and assign them to the native argument.
+ *
+ * @return jerry undefined: the validator passes,
+ *         jerry error: the validator fails.
+ */
+jerry_value_t
+jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p, /**< points to the js argument */
+                                 const jerry_length_t js_arg_cnt, /**< the count of the js arguments */
+                                 const jerryx_arg_t *c_arg_p, /**< points to the jerryx_arg_t */
+                                 const jerry_length_t c_arg_cnt) /**< the count of the native argument */
+{
+  jerry_value_t ret = jerry_create_undefined ();
+
+  if (c_arg_cnt == 0)
+  {
+    return ret;
+  }
+
+  jerryx_arg_iterator_t iterator =
```

Comment:
@martijnthe I'm usually in for future-proofness. But let me argue here (and not only for the sake of argument)

> js_arg_cnt for example is only relevant to transforms that process multiple jerry_value_ts

That's not correct. Whenever you try to get a value from the `js_arg_p` array, you need to check whether the index you try to access is valid (smaller than `js_arg_cnt`). Otherwise you have to get a newly constructed `undefined` value. And this is not specific to my approach (`jerryx_arg_helper_get `), even `jerryx_arg_iterator_{pop,peak}` work this way. That's why the iterator has to maintain both `js_arg_cnt` _and_ `js_arg_index`. The only thing that's not relevant to all transforms is `extra_info`, and that's not passed as an argument but available via `c_arg_p` only.

> if we want to enable more complex behavior (as future API or for customization inside a project), it's easy to add more iterator accessors (no API breakage)

I'd like to pick your brain what's envisioned. If the iterator is to be an opaque type then an external project will have no info on its structure and will not be able to extend its functionality. Could you please give an example on what is (or will be, or can be) possible with iterators that's not possible without them?